tests/lib.rs

```diff
@@ -876,3 +876,304 @@ Error codes:
+}
+
+#[test]
+fn redact_positional_optional_option() {
+    #[derive(FromArgs, Debug)]
+    /// Short description
+    struct Cmd {
+        #[argh(positional)]
+        /// speed of cmd
+        speed: u8,
+
+        #[argh(option)]
+        /// direction
+        direction: Option<String>,
+    }
+
+    let actual = Cmd::redact(&["<<<arg0>>>"], &["5"]).unwrap();
+    assert_eq!(actual, &["<<<arg0>>>", "speed"]);
+}
+
+#[test]
+fn redact_subcommand() {
+    #[derive(FromArgs, Debug)]
+    /// Short description
+    struct Cmd {
+        #[argh(positional)]
+        /// speed of cmd
+        speed: u8,
+
+        #[argh(subcommand)]
+        /// means of transportation
+        means: MeansSubcommand,
+    }
+
+    #[derive(FromArgs, Debug)]
+    /// Short description
+    #[argh(subcommand)]
+    enum MeansSubcommand {
+        Walking(WalkingSubcommand),
+        Biking(BikingSubcommand),
+        Driving(DrivingSubcommand),
+    }
+
+    #[derive(FromArgs, Debug)]
+    #[argh(subcommand, name = "walking")]
+    /// Short description
+    struct WalkingSubcommand {
+        #[argh(option)]
+        /// a song to listen to
+        music: String,
+    }
+
+    #[derive(FromArgs, Debug)]
+    #[argh(subcommand, name = "biking")]
+    /// Short description
+    struct BikingSubcommand {}
+    #[derive(FromArgs, Debug)]
+    #[argh(subcommand, name = "driving")]
+    /// short description
+    struct DrivingSubcommand {}
+
+    let actual = Cmd::redact(&["<<<arg0>>>"], &["5", "walking", "--music", "Bach"]).unwrap();
+    assert_eq!(actual, &["<<<arg0>>>", "speed", "walking", "--music"]);
+}
+
+#[test]
+fn redact_subcommand_with_space_in_name() {
+    #[derive(FromArgs, Debug)]
+    /// Short description
+    struct Cmd {
+        #[argh(positional)]
+        /// speed of cmd
+        speed: u8,
+
+        #[argh(subcommand)]
+        /// means of transportation
+        means: MeansSubcommand,
+    }
+
+    #[derive(FromArgs, Debug)]
+    /// Short description
+    #[argh(subcommand)]
+    enum MeansSubcommand {
+        Walking(WalkingSubcommand),
+        Biking(BikingSubcommand),
+    }
+
+    #[derive(FromArgs, Debug)]
+    #[argh(subcommand, name = "has space")]
+    /// Short description
+    struct WalkingSubcommand {
+        #[argh(option)]
+        /// a song to listen to
+        music: String,
+    }
+
+    #[derive(FromArgs, Debug)]
+    #[argh(subcommand, name = "biking")]
+    /// Short description
+    struct BikingSubcommand {}
+
+    let actual = Cmd::redact(&["<<<arg0>>>"], &["5", "has space", "--music", "Bach"]).unwrap();
+    assert_eq!(actual, &["<<<arg0>>>", "speed", "has space", "--music"]);
+}
+
+#[test]
+fn redact_produces_help() {
+    #[derive(argh::FromArgs, Debug, PartialEq)]
+    /// Woot
+    struct Repeating {
+        #[argh(option, short = 'n')]
+        /// fooey
+        n: Vec<String>,
+    }
+
+    match Repeating::redact(&["test_arg_0"], &["--help"]) {
```

Comment:
Overall, I think I struggled with this review at first because I couldn't understand what the behavior change was?

Docs will help (as described above), especially adding both examples and describing example use case(s).

But I also struggle with the name `redact`. I "get it" -- now -- but the name is more "clever" than it is descriptive, in my mind.

Since I don't think I understand the use case or purpose very well, that probably hinders my ability to come up with a "good" alternative, but maybe something along the lines of `get_args_without_values`. 

Just thinking, but would it be hard to include one more function that returns the args as provided (perhaps after parsing, so they might be normalized a little)?

If you can't think of any reason for it, then maybe it's not worth doing, but I assume someone might want to generate their own custom error messages for a given program, and embed the provided arguments. 

If you had such a function, that might make naming the redact function even more obvious (for example `get_args()` and `get_args_without_values()`, makes the second function a little easier to understand).

Reply:
> But I also struggle with the name redact. I "get it" -- now -- but the name is more "clever" than it is descriptive, in my mind.
>
> Since I don't think I understand the use case or purpose very well, that probably hinders my ability to come up with a "good" alternative, but maybe something along the lines of get_args_without_values.

I think that's fair. What do you think about the names `filter_arg_values`, or `redact_arg_values`?

> Just thinking, but would it be hard to include one more function that returns the args as provided (perhaps after parsing, so they might be normalized a little)?

That wouldn't be too hard to do, but lets wait for someone to request this first.